tokio-console/src/state/mod.rs

```diff
@@ -218,16 +229,41 @@ impl State {
 
     pub(crate) fn update_task_details(&mut self, update: proto::tasks::TaskDetails) {
         if let Some(id) = update.task_id {
-            let details = Details {
-                span_id: id.id,
-                poll_times_histogram: update.poll_times_histogram.and_then(|data| {
-                    hdrhistogram::serialization::Deserializer::new()
-                        .deserialize(&mut Cursor::new(&data))
-                        .ok()
-                }),
+            let span_id = id.id;
+            let task_id = self.tasks_state.ids.id_for(span_id);
+            let task_ref = self.tasks_state.task(task_id).unwrap();
+            let task_ref = task_ref.upgrade().unwrap();
+            let meta_id = task_ref.borrow().meta_id();
+            let poll_times_histogram = update.poll_times_histogram.and_then(|data| {
+                hdrhistogram::serialization::Deserializer::new()
+                    .deserialize(&mut Cursor::new(&data))
+                    .ok()
+            });
+            let details = self.current_task_details.take();
+            let mut fresh = Details {
+                span_id,
+                poll_times_histogram,
+                causality: details.and_then(|details| details.causality),
             };
 
-            *self.current_task_details.borrow_mut() = Some(details);
+            if let None = fresh.causality {
+                fresh.causality = Some(tracing_causality::Trace::with_root(
+                    tracing_causality::Span {
+                        id: tracing::Id::from_u64(span_id),
+                        metadata: meta_id,
+                    },
+                ));
+            }
```

Comment:
style nit, take it or leave it: this is essentially a reimplementation of [`Option::get_or_insert_with`](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.get_or_insert_with):
```suggestion
            fresh.causality.get_or_insert_with(|| tracing_causality::Trace::with_root(
                tracing_causality::Span {
                    id: tracing::Id::from_u64(span_id),
                    metadata: meta_id,
                },
            ));
```